### ISSUE TITLE
Understand how to upgrade the federation API for a resource

### ISSUE THREAD

**Issue by pmorie** [kubernetes-sigs/kubefed#804]
**What would you like to be added**:

APIs in the kubernetes ecosystem are not static and will change across releases of Kubernetes or the component that provides them. Currently, we do not have a story for how the federation API for a resource is upgraded when the API being targeted changes. This issue is for us to understand how we would support API upgrades to federation APIs.

Take for example the `Deployment` resource and a hypothetical upgrade of its API that adds a new field. At a minimum, there should be a way to mutate the federation API to include the new field.

The above example is the simplest possible case. We should also understand how we would handle a backward incompatible API changes.

**Why is this needed**:

We need a story for how to handle API upgrades in order for users of federation to keep their federation APIs up to date as they upgrade the clusters targeted by federation.

/kind feature

**Comment by pmorie** [comment]
In the April 29 WG meeting we talked about the output for this in the initial beta release being documentation that expresses what types of upgrades a user should expect to be able to handle. We expect that this area will be the subject of ongoing work as the CRD capability matures (eg conversion webhooks being elevated to beta).

**Comment by marun** [comment]
`kubefed2 enable` supports updating an existing federated type `CRD` and `FederatedTypeConfig` (FTC).  Provided backwards compatibility in the target API (i.e. no explicit conversion required), this will allow upgrading of federated types:

```bash
# Initial enable
kubefed federate enable deployments.apps --version=v1beta1

# Subsequent upgrade
kubefed federate enable deployments.apps --version=v1
```

Existing CRs for the federated type should not need to be modified so long as field changes are additive only.  However, the FTC controller does not currently watch for changes in the target version, so a running sync controller would continue to target the old endpoints until the controller manager was restarted.  I've filed #815 to track this to resolution.

**Comment by marun** [comment]
There was [an unsuccessful attempt](https://docs.google.com/document/d/1uAOMzDgczEYwKcBCuSLpdhuQPHHvRYG-w3ke9659qY0/edit#) to find a strategy to upgrade the embedded template validation schema for federated types. It was decided that the best course of action was to instead remove the embedded validation schema, which was implemented by #1038. This removes the need for kubefed to worry about how to upgrade the embedded schema of federated types.
